### PR TITLE
New extended prompt format for Canary, short utterances inference fix, and training micro-optimizations

### PR DESCRIPTION
**Pull request by pzelasko**
# What does this PR do ?

* New prompt format 
  * Prompt formatter definition
  * Special tokenizer build script 
  * Unit tests for the new Canary prompt format.
* Micro-optimizations (16% faster training step)
* Option to enable loss mask on prompt tokens
* Fix for inference of very short utterances

**Collection**: ASR

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/common/prompts/canary2.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+            prompt_template=prompt_template, expected_slots=expected_slots, slot_values=slot_values
+        )
+
+
+def map_manifest_values_to_special_tokens(slot_values: dict[str, str]) -> dict[str, str]:
+    slot_values = slot_values.copy()
+
+    any_special_token_present = False
+
+    lang_dict_compat = {"en": "en-US", "es": "es-ES", "fr": "fr-FR", "de": "de-DE"}
```
Comment (bonham79) [inline nemo/collections/common/prompts/canary2.py:113]: es-US

File: nemo/collections/common/prompts/canary2.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        # first, validate the utterance
+        expected_slots = {"source_lang", "target_lang", "pnc"}
+        missing_keys = expected_slots - set(cut.custom)
+        if missing_keys:
+            raise RuntimeError(
+                f"We found cut with ID {cut.id} that is missing the following keys: {missing_keys}"
+                f"Please ensure that every utterance in the input manifests contains these keys."
+            )
+
+        optional_slots = {
```
Comment (bonham79) [inline nemo/collections/common/prompts/canary2.py:169]: Is there a check to maintain consistency across slots or are you allowing variable slot sizes? (e.g. sample 1 has `itn` sample 2 does not have that slot.)

File: nemo/collections/common/tokenizers/canary_tokenizer.py [hunk 3]
```diff
@@ -69,22 +71,34 @@ def pad_id(self) -> int:
...
+        Tokenize the input special prompt of Canary family of models.
 
         Required because otherwise self.text_to_ids() returns a different result than what Canary had been trained with.
         """
         ans = []
-        assert text.count('>') == 5, f"Expected exactly 5 special tokens in Canary's prompt, got: {text}."
-        assert text.startswith(CANARY_BOS), text
-        for _ in range(5):
+
+        if text.startswith(CANARY2_BOCTX):
```
Comment (bonham79) [inline nemo/collections/common/tokenizers/canary_tokenizer.py:87]: Wait, can you explain this? I don't remember this in tokenizer discussions why this needs to be the case.

File: scripts/speech_recognition/canary/build_canary_1b_special_tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env python
```
Comment (bonham79) [inline scripts/speech_recognition/canary/build_canary_1b_special_tokenizer.py:1]: I don't believe this script is necessary. It can just be an example in the canary tutorial instead.

File: scripts/speech_recognition/canary/build_canary_2_special_tokenizer.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python
```
Comment (bonham79) [inline scripts/speech_recognition/canary/build_canary_2_special_tokenizer.py:1]: ditto

**Review by bonham79 — COMMENTED** [review]
So I'm a bit hesitant to create a separate Canary2 prompt class. I think it sets a bad precedent for future Canary releases that could lead to bloat. Is there any motivated reason the original Canary tokenizer class can't just redefined and then there's a separate conversion script to write up that converts the old Canary model into the new version? There's only one anyhow, so it's not like it's going to create absurd cascade problems for users.

File: nemo/collections/common/prompts/canary2.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+                    "nodiarize",
+                    "<|diarize|>",
+                    "<|nodiarize|>",
+                ),
+            },
+        },
+        # User prompt.
+        # This role is used for emotion / LID inference only - use it for two just two decoder inference steps
+        # to retrieve the recognized emotion and language tokens.
+        "user_partial": {
```
Comment (krishnacpuvvada) [inline nemo/collections/common/prompts/canary2.py:79]: Leaving a note that this doesn't cover the cases  like "identify language and transcribe with ITN" (which is fine for this version). should put sufficient guards regarding what users are allowed to ask in transcribe fn.

File: nemo/collections/common/prompts/canary2.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    for cut in cuts:
+        if isinstance(cut, MixedCut):
+            cut = cut._first_non_padding_cut
+        if not isinstance(cut, MonoCut):
+            raise TypeError(
+                f"Expected input audio to have a single channel (required MonoCut/MixedCut, but we received: {cut=})"
+            )
+
+        # first, validate the utterance
+        expected_slots = {"source_lang", "target_lang", "pnc"}
```
Comment (krishnacpuvvada) [inline nemo/collections/common/prompts/canary2.py:161]: how about `expected_slots = formatter.expected_slots` 

should we make all the slots expected ie. source_lang, target_lang, pnc, itn, timestamp, diarize and we will set the default tags in data config yaml if something is missing in manifest file ?

File: nemo/collections/common/prompts/canary2.py (hunk 2 above)
Comment (krishnacpuvvada) [reply nemo/collections/common/prompts/canary2.py:169]: we are combining different datasets annotated for different things, so the flexibility is required atleast the level of each manifest if not per sample.

File: nemo/collections/common/tokenizers/canary_tokenizer.py (hunk 3 above)
Comment (krishnacpuvvada) [reply nemo/collections/common/tokenizers/canary_tokenizer.py:87]: this is for decoding with context. the way canary currently does long-form transcripttion is to divide audio into non-overlapping chunks and run inference as independent examples. this is to be able to feed previous chunks inference into current chunks inference (and potentially explore better recognition of names if they were given in context)

File: nemo/collections/common/tokenizers/canary_tokenizer.py [hunk 8]
```diff
@@ -69,22 +71,34 @@ def pad_id(self) -> int:
...
+            # Canary 2 prompt format. It starts with decoder context, which should be tokenized using
+            # a different tokenizer than spl_tokens. We don't really know what it is, so we'll use the
+            # following HACK solution: look up 5th token which is target_lang and tokenize this part
```
Comment (krishnacpuvvada) [inline nemo/collections/common/tokenizers/canary_tokenizer.py:90]: limitation: this will error out if we are decoding with unk_lang as target_lang

File: nemo/collections/common/tokenizers/canary_tokenizer.py (hunk 3 above)
Comment (bonham79) [reply nemo/collections/common/tokenizers/canary_tokenizer.py:87]: Oh that makes way more sense

File: nemo/collections/common/prompts/canary2.py (hunk 6 above)
Comment (pzelasko) [reply nemo/collections/common/prompts/canary2.py:79]: That will be complicated to implement in transcribe fn in general. We should think about whether a new API specifically for Canary would be cleaner.

File: nemo/collections/common/prompts/canary2.py (hunk 2 above)
Comment (tbartley94) [reply nemo/collections/common/prompts/canary2.py:169]: Hmm, most of these are binary prompts. What about just having default values for the prompts instead? Else I believe y'all are adding the extra difficulty of the model needing to learn variable length prompt input.

File: nemo/collections/common/prompts/canary2.py (hunk 6 above)
Comment (tbartley94) [reply nemo/collections/common/prompts/canary2.py:79]: +1 to new API, even already the prompt setup can be a bit bulky. Putting thought into API changes before it an actual problem could save some heartache.

File: scripts/speech_recognition/canary/build_canary_1b_special_tokenizer.py (hunk 4 above)
Comment (pzelasko) [reply scripts/speech_recognition/canary/build_canary_1b_special_tokenizer.py:1]: The people building the new tokenizer for Canary 3 or community Canary forks will have an easier time and won't need to guess how the previous special tokenizers were built :)

Comment (tbartley94) [reply scripts/speech_recognition/canary/build_canary_1b_special_tokenizer.py:1]: Those people will be using the tutorial anyhow no? From my understanding, Canary use hasn't picked up that much this year, so there's not a significant community. It's more likely you'll have greater growth with the current model, so providing backend support may not be crucial.

File: nemo/collections/common/prompts/canary2.py (hunk 7 above)
Comment (pzelasko) [reply nemo/collections/common/prompts/canary2.py:161]: it will require rebuilding manifests (or at least data configs) for existing data, a bit of extra headache that I'm not sure is worth the time. WDYT?

File: nemo/collections/common/prompts/canary2.py (hunk 2 above)
Comment (pzelasko) [reply nemo/collections/common/prompts/canary2.py:169]: There is no variable length (except if you add decoder context). We indeed have default values for slots here.

File: nemo/collections/common/tokenizers/canary_tokenizer.py (hunk 8 above)
Comment (pzelasko) [reply nemo/collections/common/tokenizers/canary_tokenizer.py:90]: Yes. Better use single tokenizer in these cases...

File: scripts/speech_recognition/canary/build_canary_1b_special_tokenizer.py (hunk 4 above)
Comment (pzelasko) [reply scripts/speech_recognition/canary/build_canary_1b_special_tokenizer.py:1]: I'm re-doing the Canary 2 tokenizer right now and am grateful I put these scripts in here. Not removing them :)

File: nemo/collections/asr/modules/transformer/transformer_modules.py [hunk 9]
```diff
@@ -58,27 +58,7 @@ def _build_pos_enc(self, hidden_size, max_sequence_length, device=None):
         self.register_buffer('pos_enc', pos_enc)
 
     def forward(self, position_ids):
-        max_pos_id = position_ids.max()
-        # update positional encoding if needed
-        if max_pos_id >= self._max_sequence_length:
```
Comment (pzelasko) [inline nemo/collections/asr/modules/transformer/transformer_modules.py:63]: This check is super costly as it triggers a DtoH transfer and CUDA sync on every call to transformer decoder forward, and the proposed solution doesn't work anyway.

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 10]
```diff
@@ -67,7 +67,8 @@
 
 def lens_to_mask(lens, max_length):
     batch_size = lens.shape[0]
-    mask = torch.arange(max_length).repeat(batch_size, 1).to(lens.device) < lens[:, None]
+    arange = torch.arange(max_length, device=lens.device)
+    mask = arange.expand(batch_size, max_length) < lens.unsqueeze(1)
```
Comment (pzelasko) [inline nemo/collections/asr/models/aed_multitask_models.py:71]: micro-optimization, removes some copies and memory movement

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 11]
```diff
@@ -673,24 +674,33 @@ def training_step(self, batch: PromptedAudioToTextMiniBatch, batch_nb):
             return torch.tensor([0.0])
 
         input_ids, labels = batch.get_decoder_inputs_outputs()
+        input_ids_lens = batch.prompted_transcript_lens - 1
```
Comment (pzelasko) [inline nemo/collections/asr/models/aed_multitask_models.py:677]: fixing off-by-one issue that included an extra padding frame in decoder masks

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 12]
```diff
@@ -673,24 +674,33 @@ def training_step(self, batch: PromptedAudioToTextMiniBatch, batch_nb):
...
+
+        num_frames = batch.audio_lens.sum().float()
+        num_tokens = batch.prompted_transcript_lens.sum().float()
+        tot_frames = torch.as_tensor(batch.audio.numel(), device=num_frames.device, dtype=torch.float)
+        tot_tokens = torch.as_tensor(batch.prompted_transcript.numel(), device=num_frames.device, dtype=torch.float)
```
Comment (pzelasko) [inline nemo/collections/asr/models/aed_multitask_models.py:682]: micro-optimizations

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 13]
```diff
@@ -673,24 +674,33 @@ def training_step(self, batch: PromptedAudioToTextMiniBatch, batch_nb):
...
-        num_frames = batch.audio_lens.sum()
-        num_tokens = batch.prompted_transcript_lens.sum()
-        tot_frames = batch.audio.numel()
-        tot_tokens = batch.prompted_transcript.numel()
         tensorboard_logs = {
             'train_loss': audio_loss,
-            'learning_rate': self._optimizer.param_groups[0]['lr'],
-            'batch_size': batch.audio.shape[0],
+            'learning_rate': torch.as_tensor(self._optimizer.param_groups[0]['lr']),
+            'batch_size': torch.as_tensor(batch.audio.shape[0]),
```
Comment (pzelasko) [inline nemo/collections/asr/models/aed_multitask_models.py:703]: micro-optimizations (the PTL logger turned out to have an inefficient way of converting scalars to tensors)

File: nemo/collections/common/prompts/canary2.py (hunk 7 above)
Comment (krishnacpuvvada) [reply nemo/collections/common/prompts/canary2.py:169]: it shouldn't require us to rebuild manifests. it should be doable using the `tags` we have in train_ds yaml config? 

atleast may be the first part i.e, `expected_slots = formatter.expected_slots` ?

**Review by krishnacpuvvada — COMMENTED** [review]
LGTM overall; left a minor comment. 
Great work!

File: nemo/collections/common/prompts/canary2.py (hunk 7 above)
Comment (pzelasko) [reply nemo/collections/common/prompts/canary2.py:169]: discussed offline - keeping the current code to hide the extra complexity of new canary2 slots for users that are not interested in them

Comment (pzelasko) [reply nemo/collections/common/prompts/canary2.py]: after further conversation - made pnc optional too

**Review by krishnacpuvvada — APPROVED** [review]
LGTM
